### PR TITLE
Be consistent with rd_malloc/rd_free combos.

### PR DESCRIPTION
**Pull request by sarkanyi**
Sometimes rd_free is called on objects allocated with plain malloc. This is not consistent.

### CODE REVIEW

**Comment by sarkanyi** [comment]
And I've found some more, updating soon.

**Comment by sarkanyi** [comment]
Not sure why the unkpart test files, the commit has tehnically nothing to do with it.

**Comment by sarkanyi** [comment]
I think the failure is related to the float-linger change, but have no time to investigate.

File: src-cpp/rdkafkacpp.h [hunk 1]
```diff
@@ -81,13 +81,20 @@ typedef SSIZE_T ssize_t;
 
 /**@endcond*/
 
+#ifdef _MSC_VER    /* Visual Studio */
+#  pragma warning(disable : 4273)
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:85]: What's this for?

File: src-cpp/rdkafkacpp.h [hunk 2]
```diff
@@ -81,13 +81,20 @@ typedef SSIZE_T ssize_t;
...
+
 extern "C" {
         /* Forward declarations */
         struct rd_kafka_s;
         struct rd_kafka_topic_s;
         struct rd_kafka_message_s;
         struct rd_kafka_conf_s;
         struct rd_kafka_topic_conf_s;
+
+        RD_EXPORT void *rd_kafka_mem_malloc(struct rd_kafka_s *rk, size_t size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:96]: I'd rather see a proper C++ wrapper API, we'll want to avoid conflating the two APIs unless necessary.

File: src/lz4hc.c [hunk 3]
```diff
@@ -33,6 +33,7 @@
 */
 /* note : lz4hc is not an independent module, it requires lz4.h/lz4.c for proper compilation */
 
+#include "rd.h"   /* rd_malloc, rd_calloc, rd_free */
```
Comment (edenhill) [inline src/lz4hc.c:36]: Why is this needed?

File: src/rdkafka.h [hunk 4]
```diff
@@ -3088,6 +3088,24 @@ rd_kafka_offsets_for_times (rd_kafka_t *rk,
                             int timeout_ms);
 
 
+/**
+ * @brief Allocate memory using the librdkafka internal allocator
```
Comment (edenhill) [inline src/rdkafka.h:3092]: "using the same allocator librdkafka uses." is more correct, it currently implies that there's an internal librdkafka allocator which there is not.

File: src/rdkafka.h [hunk 5]
```diff
@@ -3088,6 +3088,24 @@ rd_kafka_offsets_for_times (rd_kafka_t *rk,
...
+ *
+ * This is typically an abstraction for the malloc(3) call and makes sure
+ * the application can use the same memory allocator as librdkafka for
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p rk must be set for memory returned by APIs that take an \c rk argument,
```
Comment (edenhill) [inline src/rdkafka.h:3098]: This does not apply to malloc, only free.

File: src/rdkafka.h [hunk 6]
```diff
@@ -3088,6 +3088,24 @@ rd_kafka_offsets_for_times (rd_kafka_t *rk,
...
+ * for other APIs pass NULL for \p rk.
+ *
+ * @remark Memory allocated by rd_kafka_mem_alloc() must be freed using
```
Comment (edenhill) [inline src/rdkafka.h:3101]: mem_malloc

**Review by edenhill — CHANGES_REQUESTED** [review]

File: src-cpp/rdkafkacpp.h (hunk 1 above)
Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h:85]: To avoid this warning:
rdkafka_example.obj : error LNK2019: unresolved external symbol __imp_rd_kafka_mem_malloc referenced in function "public: __cdecl RdKafka::Headers::Header::Header(class RdKafka::Headers::Header const &)" (??0Header@Headers@RdKafka@@QEAA@AEBV012@@Z) [C:\projects\librdkafka\win32\rdkafka_example\rdkafka_example.vcxproj]
    13>rdkafka_example.obj : error LNK2019: unresolved external symbol __imp_rd_kafka_mem_free referenced in function "void __cdecl std::_Destroy_range<struct std::_Wrap_alloc<class std::allocator<class RdKafka::Headers::Header> > >(class RdKafka::Headers::Header *,class RdKafka::Headers::Header *,struct std::_Wrap_alloc<class std::allocator<class RdKafka::Headers::Header> > &)" (??$_Destroy_range@U?$_Wrap_alloc@V?$allocator@VHeader@Headers@RdKafka@@@std@@@std@@@std@@YAXPEAVHeader@Headers@RdKafka@@0AEAU?$_Wrap_alloc@V?$allocator@VHeader@Headers@RdKafka@@@std@@@0@@Z) [C:\projects\librdkafka\win32\rdkafka_example\rdkafka_example.vcxproj]
    13>

File: src-cpp/rdkafkacpp.h [hunk 7]
```diff
@@ -146,6 +146,40 @@ std::string get_debug_contexts();
...
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p rk can be set to return memory allocated by a specific \c rk instance
+ * otherwise pass NULL for \p rk.
+ *
+ * @remark Memory allocated by mem_malloc() must be freed using
+ *         mem_free()
+ */
+RD_EXPORT
+void *mem_malloc(struct rd_kafka_s *rk, size_t size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:163]: I think it should take an optional Handle pointer, not an rk, no need to expose the C type here.

File: src-cpp/rdkafkacpp_int.h [hunk 8]
```diff
@@ -964,7 +964,7 @@ class HandleImpl : virtual public Handle {
                                    std::string &errstr) {
           char errbuf[512];
           ErrorCode err;
-          const char **extensions_copy = new const char *[extensions.size()];
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h:967]: This was actually a bug, new:ed memory should be freed with delete, not free().
Since this memory is not related to librdkafka and only lives for this call I think we should use the original new here, and change the free to delete;

File: src/rd.h [hunk 9]
```diff
@@ -154,7 +154,7 @@ static RD_INLINE RD_UNUSED char *rd_strndup(const char *s, size_t len) {
 	char *n = strndup(s, len);
 	rd_assert(n);
 #else
-	char *n = (char *)malloc(len + 1);
+	char *n = (char *)rd_malloc(len + 1);
```
Comment (edenhill) [inline src/rd.h:157]: This isn't great because if the system has strndup() we might be rd_free() that memory with another allocator (if plugged).

File: src-cpp/rdkafkacpp_int.h (hunk 8 above)
Comment (sarkanyi) [reply src-cpp/rdkafkacpp_int.h:967]: Well, since it's passed to the C function it makes sense to be malloc-ed, I guess, and free was called anyway, so in that case delete has to be called.

File: src/rd.h (hunk 9 above)
Comment (sarkanyi) [reply src/rd.h:157]: I was thinking to leave that to the other PR where you can set a different allocator. But calling system malloc is not really an option then, should call into however rd_malloc is called. But normally yes, all function should use the same allocator/free

File: src/rdkafka.h (hunk 5 above)
Comment (edenhill) [reply src/rdkafka.h]: If we have the Handle* argument there we're free to start making use of it in the future, without asking users to fix their code.

Comment (edenhill) [reply src/rdkafka.h]: .. which gives us more options to work with in the future.

File: src-cpp/rdkafkacpp_int.h (hunk 8 above)
Comment (edenhill) [reply src-cpp/rdkafkacpp_int.h:967]: The allocator doesn't matter here since the memory will never be transfered to the application or librdkafka, it is just some temporary buffer, so let's use new/delete.

File: src/rd.h (hunk 9 above)
Comment (edenhill) [reply src/rd.h:157]: Right, we can fix it in the future :+1:

File: src-cpp/rdkafkacpp.h [hunk 10]
```diff
@@ -146,6 +146,40 @@ std::string get_debug_contexts();
...
 int          wait_destroyed(int timeout_ms);
 
+/**
+ * @brief Allocate memory using the same allocator librdkafka uses.
+ *
+ * This is typically an abstraction for the malloc(3) call and makes sure
+ * the application can use the same memory allocator as librdkafka for
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p handle of type rd_kafka_t can be set to return memory allocated by a specific \c handle instance
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:156]: Why pass the C handle and not a C++ handle?

File: src/lz4.c [hunk 11]
```diff
@@ -198,10 +198,13 @@ void  LZ4_free(void* p);
 # define ALLOC_AND_ZERO(s) LZ4_calloc(1,s)
 # define FREEMEM(p)        LZ4_free(p)
 #else
-# include <stdlib.h>   /* malloc, calloc, free */
-# define ALLOC(s)          malloc(s)
-# define ALLOC_AND_ZERO(s) calloc(1,s)
-# define FREEMEM(p)        free(p)
+struct rdkafka_s;
```
Comment (edenhill) [inline src/lz4.c:201]: Why is this different from lz4frame.c ?

File: src/rdkafka.h [hunk 12]
```diff
@@ -3091,6 +3091,43 @@ rd_kafka_offsets_for_times (rd_kafka_t *rk,
                             int timeout_ms);
 
 
+
+/**
+ * @brief Allocate and zero memory using the same allocator librdkafka uses.
+ *
+ * This is typically an abstraction for the malloc(3) call and makes sure
```
Comment (edenhill) [inline src/rdkafka.h:3098]: calloc

File: src/rdkafka.h [hunk 13]
```diff
@@ -3091,6 +3091,43 @@ rd_kafka_offsets_for_times (rd_kafka_t *rk,
...
+ * the application can use the same memory allocator as librdkafka for
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p rk can be set to return memory allocated by a specific \c rk instance
+ * otherwise pass NULL for \p rk.
+ *
+ * @remark Memory allocated by rd_kafka_mem_malloc() must be freed using
```
Comment (edenhill) [inline src/rdkafka.h:3105]: calloc

**Review by edenhill — CHANGES_REQUESTED** [review]

File: src-cpp/rdkafkacpp.h (hunk 10 above)
Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h:156]: There is no C++ handle for rd_kafka_t, or maybe I missed it.

File: src/lz4.c (hunk 11 above)
Comment (sarkanyi) [reply src/lz4.c:201]: I don't exactly remember, but I think in this case we couldn't include "rd.h"

Comment (edenhill) [reply src/lz4.c:201]: It'd be good if we could use rd.h here

File: src-cpp/rdkafkacpp.h (hunk 10 above)
Comment (edenhill) [reply src-cpp/rdkafkacpp.h:156]: The RdKafka::Handle has a rk_ pointer.

File: src/lz4.c (hunk 11 above)
Comment (sarkanyi) [reply src/lz4.c:201]: lz4.h is included only internally in the lz4 code, and I think some flags differ. Including either rd.h or rdkafka_int.h will open a huge can of worms (compiler errors)

File: src-cpp/rdkafkacpp.h (hunk 10 above)
Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h]: Ok, replacing with RdKafka::Handle instance

File: src-cpp/RdKafka.cpp [hunk 14]
```diff
@@ -50,3 +50,12 @@ int RdKafka::wait_destroyed (int timeout_ms) {
   return rd_kafka_wait_destroyed(timeout_ms);
 }
 
+void *RdKafka::mem_malloc (void *handle, size_t size) {
```
Comment (edenhill) [inline src-cpp/RdKafka.cpp:53]: Use RdKafka::Handle *handle as type, not void.

File: src-cpp/rdkafkacpp.h [hunk 15]
```diff
@@ -146,6 +146,40 @@ std::string get_debug_contexts();
...
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p handle of type RdKafka::Handle can be set to return memory allocated by a
+ * specific \c handle instance otherwise pass NULL for \p handle.
+ *
+ * @remark Memory allocated by mem_malloc() must be freed using
+ *         mem_free()
+ */
+RD_EXPORT
+void *mem_malloc(void *handle, size_t size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:163]: style-nit: space before paren

File: src-cpp/rdkafkacpp.h [hunk 16]
```diff
@@ -146,6 +146,40 @@ std::string get_debug_contexts();
...
+ *
+ * This is typically an abstraction for the malloc(3) call and makes sure
+ * the application can use the same memory allocator as librdkafka for
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p handle of type RdKafka::Handle can be set to return memory allocated by a
+ * specific \c handle instance otherwise pass NULL for \p handle.
+ *
+ * @remark Memory allocated by mem_malloc() must be freed using
+ *         mem_free()
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:160]: Add trailing .

**Review by edenhill — CHANGES_REQUESTED** [review]

File: src-cpp/RdKafka.cpp (hunk 14 above)
Comment (sarkanyi) [reply src-cpp/RdKafka.cpp:53]: Ok, but then you don't mind if I forward declare "Handle" in the header.

Comment (edenhill) [reply src-cpp/RdKafka.cpp]: Maybe it is better with a mem_malloc() method on the Handle class, and one without a handle in the RdKafka namespace.

Comment (sarkanyi) [reply src-cpp/RdKafka.cpp]: Without a handle completely? So we'd have:
void *RdKafka::mem_malloc(size_t size) and
void *mem_malloc(RdKafka:Handle *handle, size_t size)?
The C api takes a handle (unused at the moment, but still)

Comment (edenhill) [reply src-cpp/RdKafka.cpp]: Yep, that was the idea.

Comment (sarkanyi) [reply src-cpp/RdKafka.cpp]: Actually I can rearrange a bit I think so there would be no need for two allocators

Comment (sarkanyi) [reply src-cpp/RdKafka.cpp]: One change though I think I should make, let it accept a NULL handle...

Comment (edenhill) [reply src-cpp/RdKafka.cpp]: My point is that there should be malloc on the Handle, and one without a handle in the global namespace, covering both scenarios. We should avoid NULL-as-optional in C++ if we can.

Comment (sarkanyi) [reply src-cpp/RdKafka.cpp]: Ok, updated with two methods (with and w/o handle)

File: src-cpp/rdkafkacpp.h [hunk 17]
```diff
@@ -3489,4 +3518,47 @@ class Metadata {
...
+ * allocating pointers that are used by librdkafka.
+ *
+ * \p handle of type RdKafka::Handle can be set to return memory allocated by a
+ * specific \c handle instance otherwise pass NULL for \p handle.
+ *
+ * @remark Memory allocated by mem_malloc() must be freed using
+ *         mem_free().
+ */
+RD_EXPORT
+void *mem_malloc (RdKafka::Handle *handle, size_t size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h:3541]: I was thinking that these two are put as instance methods on the Handle class so that the Handle argument is not needed.

Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h:3541]: That still wouldn't solve the NULL handle case. Since for example Headers has no handle.

Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h:3541]: Or should I keep the void *RdKafka::mem_malloc (size_t size); version around for just these cases?

Comment (sarkanyi) [reply src-cpp/rdkafkacpp.h:3541]: Also my guess is that if I'll include it in Handle, then you want them pure virtual

**Review by edenhill — APPROVED** [review]
LGTM!

**Comment by edenhill** [comment]
Thanks for your perseverence!
